pkg/controllers/resources/persistentvolumeclaims/translate.go

```diff
@@ -24,12 +24,19 @@ func (s *persistentVolumeClaimSyncer) translate(ctx *synccontext.SyncContext, vP
 	if err != nil {
 		return nil, err
 	}
-	if newPvc.Spec.DataSource != nil && vPvc.Annotations[constants.SkipTranslationAnnotation] != "true" &&
-		(newPvc.Spec.DataSource.Kind == "PersistentVolumeClaim" || newPvc.Spec.DataSource.Kind == "VolumeSnapshot") {
-		newPvc.Spec.DataSource.Name = translate.Default.PhysicalName(newPvc.Spec.DataSource.Name, vPvc.Namespace)
+
+	if vPvc.Annotations[constants.SkipTranslationAnnotation] != "true" {
+		if newPvc.Spec.DataSource != nil &&
+			(newPvc.Spec.DataSource.Kind == "PersistentVolumeClaim" || newPvc.Spec.DataSource.Kind == "VolumeSnapshot") {
+			newPvc.Spec.DataSource.Name = translate.Default.PhysicalName(newPvc.Spec.DataSource.Name, vPvc.Namespace)
+		}
+
+		if newPvc.Spec.DataSourceRef != nil &&
+			(newPvc.Spec.DataSourceRef.Kind == "PersistentVolumeClaim" || newPvc.Spec.DataSourceRef.Kind == "VolumeSnapshot") {
```

Comment:
I don't think we should check Kind here^ because DataSourceRef supports arbitrary Kind. But because "The API server ensures that the two fields always have the same contents", we should not check Kind of the DataSource either. 
Relevant docs - https://kubernetes.io/blog/2021/08/30/volume-populators-redesigned/

Reply:
Yes that's correct. Since, the API Server ensures the proper KIND(s) for the dataSource and dataSourceRef fields while creating the resource itself, we don't need to check the KIND again while translating. I have made the necessary changes and pushed the updated code. Thanks @matskiv .